### PR TITLE
WIP : Logging: Add File Logging

### PR DESCRIPTION
**Pull request by nemaniarjun**
Running `coala` in all modes except `json`
will output a log file in the folder it has been run
named `coala-log` with all log messages.

Closes https://github.com/coala/coala/issues/1862

### Checklist

- [ ] I have rebased properly. Please see [our tutorial on rebasing](http://coala.io/git#rebasing).
- [ ] I have gone through the [commit guidelines](http://coala.io/commit) and I've followed them.
- [ ] I have followed the [guidelines for the commit shortlog](http://coala.io/commit#shortlog).
- [ ] I have followed the [guidelines for the commit body](http://coala.io/commit#commit-body).
- [ ] I have included the issue URL with the 'Fixes' keyword if the commit fixes a *real bug* and the 'Closes' keyword if it adds a feature or enhancement. For details, check the [issue reference section in our commit guidelines](http://coala.io/commit#issue-reference).
- [ ] I have [run all the tests](http://api.coala.io/en/latest/Developers/Executing_Tests.html) and they all pass. If any CI (below) is not green, please fix them. If GitMate issues appear you will have to amend your commits, pushing new commits on top is not sufficient!

### CODE REVIEW

**Comment by nemaniarjun** [comment]
CC @Makman2

File: .gitignore [hunk 1]
```diff
@@ -36,3 +36,4 @@ node_modules
 .vendor
 Goopfile.lock
 venv/
+coala-log
```
Comment (Makman2) [inline .gitignore]: miep

File: coalib/output/Logging.py [hunk 2]
```diff
@@ -15,18 +15,24 @@ def configure_logging():
     logging.config.dictConfig({
         'version': 1,
         'handlers': {
-            'colored': {
+            'console-logger': {
```
Comment (Makman2) [inline coalib/output/Logging.py]: I think renaming the handler should go into an own commit^^

File: coalib/output/Logging.py [hunk 3]
```diff
@@ -15,18 +15,24 @@ def configure_logging():
...
                 'class': 'logging.StreamHandler',
-                'formatter': 'colored',
+                'formatter': 'color',
```
Comment (Makman2) [inline coalib/output/Logging.py]: why changing this setting suddenly?

File: coalib/output/Logging.py [hunk 4]
```diff
@@ -15,18 +15,24 @@ def configure_logging():
...
                 'stream': sys.stderr
+            },
+            'file-logger': {
+                'class': 'logging.FileHandler',
+                'formatter': 'no-color',
+                'filename': 'coala-log',
+                'mode': 'w+'
```
Comment (Makman2) [inline coalib/output/Logging.py]: I would truncate, otherwise the log will grow quite big^^

Comment (Makman2) [inline coalib/output/Logging.py]: Also can you turn on debug log messages? That's the whole purpose of the log file actually ;)

File: coalib/output/Logging.py [hunk 5]
```diff
@@ -37,6 +43,12 @@ def configure_logging():
                     'INFO': 'blue',
                     'DEBUG': 'green'
                 }
+            },
+            'no-color': {
+                '()': 'logging.Formatter',
+                'format': '[%(levelname)s][%(asctime)s]'
+                          ' %(message)s',
```
Comment (Makman2) [inline coalib/output/Logging.py]: doesn't fit into a single line? :)

File: coalib/output/Logging.py (hunk 3 above)
Comment (nemaniarjun) [inline coalib/output/Logging.py]: Just for the sake of it. Color and no-color, instead of colored and non-colored. 
This setting doesn't have any effect outside this function. So it doesn't matter much.

Comment (Makman2) [inline coalib/output/Logging.py]: then -> own commit ;)

File: coalib/output/Logging.py (hunk 4 above)
Comment (nemaniarjun) [inline coalib/output/Logging.py]: What do you mean by truncate? 
> `w+` Opens a file for both writing and reading. Overwrites the existing file if the file exists. If the file does not exist, creates a new file for reading and writing.

File: coalib/output/Logging.py (hunk 3 above)
Comment (nemaniarjun) [inline coalib/output/Logging.py]: One commit for renaming both handler and formatter seems fair

File: coalib/output/Logging.py (hunk 4 above)
Comment (Makman2) [inline coalib/output/Logging.py]: truncating a file means that you remove its contents on open.
seems `w+` does truncate ;) but do you need `+`?

Comment (nemaniarjun) [inline coalib/output/Logging.py]: I am having a bit of difficulty doing this. I can't understand how the arg `-L ` works.

Comment (Makman2) [inline coalib/output/Logging.py]: ok then let's do this later :+1:

Comment (nemaniarjun) [inline coalib/output/Logging.py]: No I dont think so :)

Comment (nemaniarjun) [inline coalib/output/Logging.py]: I am trying to understand this :')

**Comment by nemaniarjun** [comment]
This is with respect to why the build failed.

Got a problem, in coalaDeleteOrigTest under `test_nonexistent_coafile` we are mocking `os.getcwd()` to be `None`. This create a problem when I try to create a log file. (since that would need an actual path to be made.)
Should I change this to point to a empty folder? (or maybe create and delete one for this test?)


another fix : configure_logging after the above test ends. i.e. on L22 instead of L14 in https://github.com/coala/coala/blob/master/coalib/coala_delete_orig.py#L14

File: coalib/output/Logging.py (hunk 4 above)
Comment (nemaniarjun) [inline coalib/output/Logging.py]: Nearly done. Requires a certain amount of testing, if you have any corner cases that need to checked do comment :)

**Comment by nemaniarjun** [comment]
@Makman2 according to your solution I tried the first way, but that wont work as it will require changing some lines in `ConfigurationGathering` something i am unwilling to do.

Also 3rd Fix will not work (It will just try to config logging again)

Thought of a 4th Fix: change the name of the current `main` function to something else and test that.
while configuring logging in a new main function.

**Comment by Makman2** [comment]
@nemaniarjun you are talking about using a temp-directory for tests, right?

> but that wont work as it will require changing some lines in ConfigurationGathering something i am unwilling to do.

what exactly needs change?

> Also 3rd Fix will not work (It will just try to config logging again)

what 3rd fix?

> Thought of a 4th Fix: change the name of the current main function to something else and test that.
while configuring logging in a new main function.

okay don't exactly understand this, just try ;)

**Comment by nemaniarjun** [comment]
I tried to import `args` from `coalib/coala.py` and use it to check if the mode is json in `Logging.py` but I get a ImportError

**Comment by Makman2** [comment]
So if I understand correctly, json-logging-configuration overrides the configuration that contains the file-handler?

**Comment by nemaniarjun** [comment]
Nope! When run in json mode, first I override with json-logging-config, but after that when log_printer is used it overwrites the json-logging with file-logging

**Comment by Makman2** [comment]
Ah yeah sorry confused the order :P :P

Imo there are two possibilities:
- We merge all logging configuration functions into a single one, so we can handle them correctly together
- We modularize json-logging, normal stdout logging and file logging (first two are already modularized), and calling one of the functions does not override other configurations.

I prefer 2nd if possible (so we have then `enable_json_logging`, `enable_file_logging`, `enable_std_logging` or so, together with a `reset_logging` function). @nemaniarjun is it possible to patch the logging configuration once it's set?

**Comment by nemaniarjun** [comment]
Found [this](https://docs.python.org/3/library/logging.config.html#incremental-configuration)

**Comment by Makman2** [comment]
Looks like the right direction, though this worries me:

> Thus, when the incremental key of a configuration dict is present and is True, the system will completely ignore any formatters and filters entries, and process only the level settings in the handlers entries, and the level and propagate settings in the loggers and root entries.

Do we get problems with this behaviour?

**Comment by nemaniarjun** [comment]
I think this is good. And I can figure this out. It might cause a problem but got an intuition that it will not be a big one.

**Comment by Makman2** [comment]
@nemaniarjun cool :)

**Comment by nemaniarjun** [comment]
@Makman2 This works :D 
Can you however checkout this branch and try it once on your computer ?
Thanks :)

**Comment by nemaniarjun** [comment]
Don't know why the test failed, some error in gitmate?

File: .gitignore [hunk 6]
```diff
@@ -36,3 +36,4 @@ node_modules
 .vendor
 Goopfile.lock
 venv/
+**/coala-log
```
Comment (Makman2) [inline .gitignore]: you don't need `**` :)

File: coalib/coala_delete_orig.py [hunk 7]
```diff
@@ -13,7 +13,10 @@
 
 def main(section: Section=None):
     configure_logging()
+    return start_delete_orig(section)
 
+
+def start_delete_orig(section: Section=None):
```
Comment (Makman2) [inline coalib/coala_delete_orig.py]: good change, but separate commit :+1:

File: coalib/coala_delete_orig.py [hunk 8]
```diff
@@ -26,7 +29,7 @@ def main(section: Section=None):
     not_deleted = 0
     for ofile in orig_files:
         logging.info('Deleting old backup file... '
-                         + os.path.relpath(ofile))
+                     + os.path.relpath(ofile))
```
Comment (Makman2) [inline coalib/coala_delete_orig.py]: actually I'm not sure about this one...
1. definitely separate commit (also for the change below)
2. could you read up in PEP8 again how this shall be handled? The problem is we seem to allow both styles, and next time some guy changes it back^^ If PEP8 has no opinion on this case, I would rather tell you to revert this change, sorry :/

File: coalib/misc/Caching.py [hunk 9]
```diff
@@ -14,8 +14,11 @@ class FileCache:
 
     >>> from pyprint.NullPrinter import NullPrinter
     >>> from coalib.output.printers.LogPrinter import LogPrinter
+    >>> from coalib.output.Logging import configure_logging
     >>> import logging
     >>> import copy, time
+    >>> configure_logging()
+    20
```
Comment (Makman2) [inline coalib/misc/Caching.py]: what is `20`? :O

File: coalib/misc/Caching.py [hunk 10]
```diff
@@ -14,8 +14,11 @@ class FileCache:
...
     >>> log_printer = LogPrinter()
     >>> log_printer.log_level = logging.CRITICAL
```
Comment (Makman2) [inline coalib/misc/Caching.py]: I think you can omit those then

Comment (Makman2) [inline coalib/misc/Caching.py]: Nvm

Comment (Makman2) [inline coalib/misc/Caching.py]: Maybe separate commit?

Comment (Makman2) [inline coalib/misc/Caching.py]: We definitely need to explain why we `configure_logging` and then use a `LogPrinter`, this is definitely confusing for unaware people^^

File: coalib/output/Logging.py [hunk 11]
```diff
@@ -4,25 +4,40 @@
...
+
+def configure_logging(new_log_level=logging.INFO, incremental=False):
     """
     Configures the logging with hard coded dictionary.
     """
     import sys
 
     logging.config.dictConfig({
         'version': 1,
+        'incremental': incremental,
```
Comment (Makman2) [inline coalib/output/Logging.py]: Don't we want to have this set to `True`?

File: coalib/output/Logging.py [hunk 12]
```diff
@@ -4,25 +4,40 @@
...
                 'formatter': 'color',
-                'stream': sys.stderr
+                'stream': sys.stderr,
+                'level': LOG_LEVEL.reverse.get(new_log_level)
+            },
+            'file-handler': {
+                'class': 'logging.FileHandler',
+                'formatter': 'no-color',
+                'filename': 'coala-log',
+                'mode': 'w+',
```
Comment (Makman2) [inline coalib/output/Logging.py]: shouldn't this be just `w`?

File: coalib/output/printers/LogPrinter.py [hunk 13]
```diff
@@ -121,22 +122,26 @@ def __init__(self,
...
+        if not printer:
+            self.log_level = log_level
 
     @property
     def log_level(self):
         """
         Returns current log_level used in logger.
         """
-        return self.logger.getEffectiveLevel()
+        # return self.logger.getEffectiveLevel()
```
Comment (Makman2) [inline coalib/output/printers/LogPrinter.py]: delete commented out code ;)

File: coalib/output/printers/LogPrinter.py [hunk 14]
```diff
@@ -121,22 +122,26 @@ def __init__(self,
...
+        return self.level
 
     @log_level.setter
     def log_level(self, log_level):
         """
         Sets log_level for logger.
         """
-        self.logger.setLevel(log_level)
+        # self.logger.setLevel(log_level)
+        # print(self.printer)
```
Comment (Makman2) [inline coalib/output/printers/LogPrinter.py]: commented out code

File: coalib/output/Logging.py (hunk 11 above)
Comment (nemaniarjun) [inline coalib/output/Logging.py]: Not always, thats why taking a new argument, which by default is False.

File: coalib/coala_delete_orig.py (hunk 8 above)
Comment (nemaniarjun) [inline coalib/coala_delete_orig.py]: I didnt make this change, it was made by PystyleCodeBear

File: .gitignore (hunk 6 above)
Comment (nemaniarjun) [inline .gitignore]: Sometimes there are extra logs created in tests folder, to ignore these too I used the `**` :D

File: coalib/misc/Caching.py (hunk 10 above)
Comment (nemaniarjun) [inline coalib/misc/Caching.py]: Should I put that comment here or in `LogPrinter.py`?

File: coalib/output/Logging.py (hunk 11 above)
Comment (nemaniarjun) [inline coalib/output/Logging.py]: When `configure_logging` is called for the first time, the increment should be `False`, when using it to update the levels we set it to `True`, thus this change.

File: coalib/output/Logging.py (hunk 12 above)
Comment (nemaniarjun) [inline coalib/output/Logging.py]: Yeah forgot to change that!

File: coalib/coala_delete_orig.py [hunk 15]
```diff
@@ -26,17 +25,17 @@ def main(log_printer=None, section: Section=None):
 
     not_deleted = 0
     for ofile in orig_files:
-        log_printer.info('Deleting old backup file... '
+        logging.info('Deleting old backup file... '
                          + os.path.relpath(ofile))
```
Comment (nemaniarjun) [inline coalib/coala_delete_orig.py]: @Makman2 This is still weird.

CC @sils because Gitmate issue.

File: coalib/coala_delete_orig.py (hunk 8 above)
Comment (Makman2) [inline coalib/coala_delete_orig.py]: okay, separate commit then :)

File: coalib/misc/Caching.py (hunk 10 above)
Comment (Makman2) [inline coalib/misc/Caching.py]: here before `log_printer = LogPrinter()`

File: coalib/output/Logging.py (hunk 11 above)
Comment (Makman2) [inline coalib/output/Logging.py]: hm this sounds complicated. When do we update levels?

File: coalib/misc/Caching.py (hunk 10 above)
Comment (nemaniarjun) [inline coalib/misc/Caching.py]: Okay, also added a comment to `LogPrinter`

File: coalib/output/Logging.py (hunk 11 above)
Comment (nemaniarjun) [inline coalib/output/Logging.py]: We use levels a lot in testing and doctests.
I dont know exactly when we do this.
But basically when we configure_logging for the first time we set increment to false.
When we only want to change the level we set it to True.

**Comment by nemaniarjun** [comment]
Creating a new PR, too many comments and gitmate is still irritating about that change :(
